types/ramda/index.d.ts

```diff
@@ -741,10 +741,7 @@ export function F(...args: unknown[]): false;
 /**
  * Returns a new list containing only those items that match a given predicate function. The predicate function is passed one argument: (value).
  */
-export function filter<A, P extends A>(pred: (val: A) => val is P): {
-    <B extends A>(list: readonly B[]): P[];
-    <B extends A>(dict: Dictionary<B>): Dictionary<P>;
-};
+// https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#use-union-types
```

Comment:
Did you forgot to reimplement this case with union type?

Reply:
After some research, I learned `is` keyword, the difference between function overloads and union type. I think it's both ok to use function overloads and union type in this situation.

Reply:
This PR is trying to solve the following problem:

```ts
  const pipeFilterEven = R.pipe(
    R.map(Boolean),
    R.filter(isEven)
  );
```

TS complain that:

```ts
Argument of type '<P extends number, C extends readonly P[] | Dictionary<P>>(collection: C) => C' is not assignable to parameter of type '(a: boolean[]) => readonly number[] | Dictionary<number>'.
  Types of parameters 'collection' and 'a' are incompatible.
    Type 'boolean[]' is not assignable to type 'readonly number[] | Dictionary<number>'.
      Type 'boolean[]' is not assignable to type 'Dictionary<number>'.
        Index signature for type 'string' is missing in type 'boolean[]'.ts(2345)
function filter<number>(pred: (value: number) => boolean): <P, C>(collection: C) => C (+4 overloads)
```

this is because curried filter return two overloads function type, and `R.pipe` don't know how to combine overload functions with the following function (`R.map(Boolean)` in the above example).

So I change it to union:

```ts
export function filter<A, P extends A>(pred: (val: A) => val is P): {
  <B extends A, C extends (B[] | Dictionary<B>), D extends (P[] | Dictionary<P>)>(list: C): D;
};
```

but this time, TS also don't know how to combine `R.filter(isEven)` with `R.map(Boolean)`, this time the reason is it can't put a union type to `R.map`.

Finally, I solve it by specifying the R.filter(isEven) input type explicitly, instead of point-free style, as following:

```ts
  const pipeFilterEvenV2 = R.pipe(
    R.map(Boolean),
    (xs: any[]) => R.filter(isEven, xs)
  );
```

not so elegant, but it works.

I'll think about how to optimize this kind of situation.

Reply:
I find conditional type may be help.